### PR TITLE
Fuzzy checking: Bayesian version

### PR DESCRIPTION
**Pull request by zmbc**
### Description
- *Category*: test
- *JIRA issue*: [SSCI-995](https://jira.ihme.washington.edu/browse/SSCI-995) (research ticket)
- *Research reference*: none

### Changes and notes

**Terminology**:
- Here, I am considering our existing integration tests to be part of "V&V" (which I write "v_and_v" in file paths). They are the automated part, while our traditional V&V process is the manual part. This PR adds automated V&V/integration testing **of proportions** (that are not 0 or 1).
- "Fuzzy checking" is the term I am using for checking anything that is subject to sampling error/stochastic variation. For the time being, this is only proportions, but could be expanded to other types of quantities in the future.

Statistically, this uses Bayesian hypothesis testing. Each check compares two models/hypotheses of the underlying rate, one that represents "if there is a bug" and one that represents "if there is not a bug." We use a simple Jeffreys beta prior for "if there is a bug." For "if there is not a bug" we fit a beta distribution to a 95% confidence interval provided by the user. In both cases, this is then fed through a binomial (creating a beta-binomial) for the actual count data. We fail the tests if any check is "decisive" (Bayes factor > 100) in favor of "there is a bug."

Technically, this uses a pytest fixture that is kept active for the entire test session and allows testing these hypotheses while maintaining a log of information about them. This diagnostic information is output at the end of a test run for optional human inspection. I have gitignored the diagnostic output -- is this the right call?

### Verification and Testing

I have verified that these tests pass. I have also verified that they do _not_ pass, even with relatively small population (20k), when I intentionally introduce bugs into the simulation. The two that I tried were:

* Multiplying all domestic migration rates by 1.5
* Always rounding the immigration rate upward instead of using stochastic rounding

### CODE REVIEW

**Comment by zmbc** [comment]
Note: I'm currently hoping to get review from RT on this math, both here and in https://github.com/ihmeuw/vivarium_research/pull/1382. Only after that do I think it makes sense for engineering folks to review this for code quality.

File: integration_tests/conftest.py
Comment (albrja) [inline integration_tests/conftest.py:254]: This is also just writing an output file to look at for further investigation to go along with the normal test suites having to pass correct?

File: integration_tests/conftest.py
Comment (albrja) [inline integration_tests/conftest.py:203]: How (or why) is this being cached? Doesn't this method have to run if you hit the branch in fuzzy_assert_proportions?

File: integration_tests/conftest.py
Comment (zmbc) [reply integration_tests/conftest.py:254]: Yes, this is an output file a human could inspect, either just for visibility into how the fuzzy checking is working overall, or to see more details about the cause of warnings. I've proposed here that we output this file each time we run the integration tests, but do _not_ track it in source control, to avoid lots of diff/merge conflicts on what is essentially a build artifact.

File: integration_tests/conftest.py
Comment (zmbc) [reply integration_tests/conftest.py:203]: This is being cached in memory, so if it is called multiple times in the same run with the same lb and ub parameters, it doesn't re-do the calculation, just uses the value it got last time.

In cases where we check something every timestep, for instance, this would prevent us from having to keep recalculating the beta for the same target UI over and over.

I didn't look at how much this cache is actually being used but it seems very low cost.

File: integration_tests/conftest.py
Comment (zmbc) [inline integration_tests/conftest.py:105]: Could use a tuple parameter here to avoid the complexity of the checks/optionals.

File: integration_tests/conftest.py
Comment (zmbc) [inline integration_tests/conftest.py:106]: Don't abbreviate `lb`, `ub` etc.

File: integration_tests/conftest.py
Comment (zmbc) [inline integration_tests/conftest.py:104]: Should we just pass in the aggregations we need? Proportion, length?

File: integration_tests/test_domestic_migration.py
Comment (zmbc) [inline integration_tests/test_domestic_migration.py:58]: This "magic value" (0.99) should have units and be in the YAML.

File: integration_tests/test_domestic_migration.py
Comment (zmbc) [inline integration_tests/test_domestic_migration.py:76]: For simplicity, let's aggregate this from the input drift per unit time, assuming each timestep has equal population size (roughly true). Create a function to do this.

**Review by zmbc — COMMENTED** [review]
One future concern will be how we deal with this in CI, once integration tests are run in CI. Rajan's idea is to have a nightly build run the integration tests, and that nightly build should do the following:
* First, run the integration tests with a seed that is chosen randomly (different every day, and different than the engineers run manually)
* If that fails, also run the tests with the "standard" seed that is used by the engineers routinely; this will allow us to separate random failures due to seed from bugs that just weren't caught before merging

Comments transcribed from Rajan (errors are my own):

File: integration_tests/conftest.py
Comment (rmudambi) [inline integration_tests/conftest.py:362]: Nit: shouldn't this be `drifts` not `bounds`? i.e.
```suggestion
def multiplicative_drifts_to_bound_at_timestep(
```

File: integration_tests/conftest.py
Comment (rmudambi) [inline integration_tests/conftest.py:390]: Same here

File: integration_tests/test_domestic_migration.py
Comment (zmbc) [inline integration_tests/test_domestic_migration.py:21]: This is where we load the YAML targets.

File: integration_tests/test_domestic_migration.py
Comment (zmbc) [inline integration_tests/test_domestic_migration.py:68]: Here's an example of a converted assert -- instead of .any(), we check the proportion using the target. Most of the diff in the test files is these conversions.

File: integration_tests/test_domestic_migration.py
Comment (zmbc) [inline integration_tests/test_domestic_migration.py:138]: Note: This is an unusual one where I think it makes sense for the target to be set on the engineering side. It can be very conservative.

File: integration_tests/test_domestic_migration.py
Comment (zmbc) [inline integration_tests/test_domestic_migration.py:239]: Note: This is an unusual one where I think setting the target UI on the engineering side makes sense. It can be very conservative. It's kind of a strange assertion (see line 73 on the deleted side).

File: integration_tests/test_domestic_migration.py
Comment (rmudambi) [inline integration_tests/test_domestic_migration.py:189]: These values are used twice (at least?). Store them as constants.

**Review by rmudambi — APPROVED** [review]

File: integration_tests/conftest.py
Comment (zmbc) [reply integration_tests/conftest.py:362]: There are two drift values input, and two bound values output. So maybe it's both -- `multiplicative_drifts_to_bounds_at_timestep`.

File: integration_tests/test_domestic_migration.py
Comment (zmbc) [reply integration_tests/test_domestic_migration.py:138]: Fair enough. This one felt a little weird to me because it's not something I can imagine the RT thinking to validate ahead of time -- it's such a corner case. If I'm reading your comment correctly, that's okay -- but at the point where engineering decided to check this, they would have asked RT to specify the value?

If I'm understanding correctly, that makes sense to me; I can throw this in a YAML on the research side.

File: integration_tests/test_domestic_migration.py
Comment (zmbc) [reply integration_tests/test_domestic_migration.py:239]: I think this is in the same category as the one above.

File: integration_tests/test_domestic_migration.py
Comment (rmudambi) [reply integration_tests/test_domestic_migration.py:138]: Yes, that's exactly what I meant!

**Review by rmudambi — APPROVED** [review]

File: integration_tests/conftest.py
Comment (patricktnast) [inline integration_tests/conftest.py:146]: I think it's mostly a matter of taste in this application, but any chance I could convince you to express the Bayes factor on a decibel scale, a la Jaynes and Good?

Comment (zmbc) [reply integration_tests/conftest.py:146]: Interesting, TIL! I am not opposed, but would prefer to postpone that to a different PR. There are a number of directions we could go here, and the one that feels most useful to me would be to specify desired sensitivity or specificity, given the prior distributions.

The intention here is that these cutoffs wouldn't be changed often. In fact, I don't think they've had to be tweaked at all here or in pseudopeople.

Comment (patricktnast) [reply integration_tests/conftest.py:146]: yeah, i'm thinking more about the output message of a particular test than how to specify the cutoffs--regardless of how you specify the cutoff, would you report the test outcome in terms of odds like the Bayes Factor against a cutoff value? If so, reporting that in terms of decibel / log odds is nice/more interpretable for the human reading it (or at least for me 😄).

w.r.t. choosing the cutoffs, have you thought about going a step further in the default case and choose boundaries that given the prior optimize some discrimination metric like F score or log loss or something? Of course, in general I think you're right that it makes sense to let the test designer prioritize sensitivity / specificity/ some confusion matrix metric.

Comment (zmbc) [reply integration_tests/conftest.py:146]: Interesting ideas. I agree the output message could be improved; right now the Bayes factor gets printed in scientific notation a lot since it can get so large.

I found [this](https://www.johndcook.com/blog/2008/03/11/how-loud-is-the-evidence/); is this the same decibel scale you are thinking of? So with this cutoff, the test would fail when the evidence was more than 20 dB. Seems like a more intuitive way to think about it.

Really interesting idea about the discrimination metric. I am thinking about this in terms of decision theory, and I think in most cases the cost of a miss will be quite asymmetric. It might be interesting for _that_ to be the configurable parameter: what is the cost of not catching an issue vs raising a false alarm? But, the more fancy we get, the more we are relying on the prior distributions being somewhat accurate.

Comment (patricktnast) [reply integration_tests/conftest.py:146]: Yes, that's right--Jaynes calls it "evidence" but it has [several names](https://en.wikipedia.org/wiki/Hartley_(unit)). The particular points in favor, one of which the post you found covers:

- It's symmetric around zero (50%), so e.g. 10 and -10 dB represent the same odds for and against
- It is additive, so it is easy to keep a running total of dB with new pieces of evidence, which proponents identify with the procedure of incremental Bayesian updates
- Folks who are familiar with working with a decibel scale may know the simple power ratio heuristic 3 dB ~ 2x power and 10 dB = 10x power
-  It is speculated to be calibrated to human perception in the vein of the [Weber Fetchner law](https://en.wikipedia.org/wiki/Weber%E2%80%93Fechner_law), so that 1 dB is about as much "evidential strength" that a person can meaningfully distinguish (~25% increase)

> I am thinking about this in terms of decision theory, and I think in most cases the cost of a miss will be quite asymmetric. It might be interesting for that to be the configurable parameter: what is the cost of not catching an issue vs raising a false alarm?

Yeah I'm not super familiar with evaluation metrics--there seem to be many, especially for binary classifiers--but just looking at wikipedia they show a [parameterized Beta f-score](https://en.wikipedia.org/wiki/F-score) that weights the harmonic mean toward precision or recall. I wonder if it would be sensible to do something similar with sensitivity / specificity. At the end of the day you're trying to trade off type I and type II errors i guess, but I'm not sure how it all pencils out. But it would be nice to say "I want to prioritize sensitivty / specificity X times more than the other" without having to know the right cutoffs

> But, the more fancy we get, the more we are relying on the prior distributions being somewhat accurate.
Yeah, this is definitely an argument not to go too crazy with the information theory stuff

Comment (zmbc) [reply integration_tests/conftest.py:146]: I believe any decision metric will correspond directly to some loss function. In the case of a binary classification, the loss function is just a 2x2 matrix of how good/bad it is to have any combination of predicted and true class. My guess is that the loss function itself will be more intuitive to specify than the corresponding metric or even the desired sensitivity/specificity.

**Review by patricktnast — APPROVED** [review]

**Pull request merged**
